plugin/trino-ranger/src/main/java/io/trino/plugin/ranger/RangerSystemAccessControl.java

```diff
@@ -0,0 +1,1391 @@
+
+        return res;
+    }
+
+    private static Set<RangerTrinoResource> createRoleResources(Set<String> roleNames)
+    {
+        Set<RangerTrinoResource> ret = new HashSet<>(roleNames.size());
+
+        for (String rolName : roleNames) {
+            ret.add(createRoleResource(rolName));
+        }
+
+        return ret;
+    }
+
+    private static RangerTrinoResource createResource(QueryId queryId)
+    {
+        RangerTrinoResource res = new RangerTrinoResource();
+
+        res.setValue(RangerTrinoResource.KEY_QUERY_ID, queryId != null ? queryId.getId() : "*");
+
+        return res;
+    }
+
+    private void setDefaultConfig(RangerConfig config)
+    {
+        if (StringUtils.isBlank(config.getServiceName())) {
+            config.setServiceName(RANGER_TRINO_DEFAULT_SERVICE_NAME);
+        }
+
+        if (config.getPluginConfigResource() == null || config.getPluginConfigResource().isEmpty()) {
+            config.setPluginConfigResource(Arrays.asList(RANGER_TRINO_DEFAULT_SECURITY_CONF, RANGER_TRINO_DEFAULT_AUDIT_CONF, RANGER_TRINO_DEFAULT_POLICYMGR_SSL_CONF));
+        }
+
+        if (config.getHadoopConfigResource() == null || config.getHadoopConfigResource().isEmpty()) {
+            config.setHadoopConfigResource(Arrays.asList(RANGER_TRINO_DEFAULT_HADOOP_CONF));
+        }
+    }
+
+    private static Identity toIdentity(Optional<Principal> principal)
+    {
+        return principal.isPresent() ? Identity.ofUser(principal.get().getName()) : Identity.ofUser("");
+    }
+
+    private static class RangerTrinoEventListener
+            implements EventListener
+    {
+        private final Map<String, QueryCreatedEvent> activeQueries = new HashMap<>();
+
+        public String getClientAddress(String queryId)
+        {
+            QueryCreatedEvent qce = activeQueries.get(queryId);
+            QueryContext qc = qce != null ? qce.getContext() : null;
+
+            return qc != null && qc.getRemoteClientAddress().isPresent() ? qc.getRemoteClientAddress().get() : null;
+        }
+
+        public String getQueryText(String queryId)
+        {
+            QueryCreatedEvent qce = activeQueries.get(queryId);
+            QueryMetadata qm = qce != null ? qce.getMetadata() : null;
+
+            return qm != null ? qm.getQuery() : null;
+        }
+
+        public Instant getQueryTime(String queryId)
+        {
+            QueryCreatedEvent qce = activeQueries.get(queryId);
+
+            return qce != null ? qce.getCreateTime() : null;
+        }
+
+        public String getClientType(String queryId)
+        {
+            QueryCreatedEvent qce = activeQueries.get(queryId);
+            QueryContext qc = qce != null ? qce.getContext() : null;
+
+            return qc != null && qc.getUserAgent().isPresent() ? qc.getUserAgent().get() : null;
+        }
+
+        @Override
+        public void queryCreated(QueryCreatedEvent queryCreatedEvent)
+        {
+            QueryMetadata qm = queryCreatedEvent.getMetadata();
```

Comment:
do not use shortcuts like `qm`. `QueryManager` should not be null. That would be a bug in Trino.